src/hpa.c

```diff
@@ -0,0 +1,395 @@
+	    WITNESS_RANK_HPA_SHARD, malloc_mutex_rank_exclusive);
+	if (err) {
+		return true;
+	}
+
+	shard->edata_cache = edata_cache;
+	shard->hpa = hpa;
+	psset_init(&shard->psset);
+	shard->ps_goal = ps_goal;
+	shard->ps_alloc_max = ps_alloc_max;
+
+	/*
+	 * Fill these in last, so that if an hpa_shard gets used despite
+	 * initialization failing, we'll at least crash instead of just
+	 * operating on corrupted data.
+	 */
+	shard->pai.alloc = &hpa_alloc;
+	shard->pai.expand = &hpa_expand;
+	shard->pai.shrink = &hpa_shrink;
+	shard->pai.dalloc = &hpa_dalloc;
+
+	shard->ind = ind;
+	assert(ind == base_ind_get(edata_cache->base));
+
+	return false;
+}
+
+static edata_t *
+hpa_alloc_central(tsdn_t *tsdn, hpa_shard_t *shard, size_t size_min,
+    size_t size_goal) {
+	bool err;
+	edata_t *edata;
+
+	hpa_t *hpa = shard->hpa;
+
+	malloc_mutex_lock(tsdn, &hpa->mtx);
+	edata = hpa_central_alloc_reuse(tsdn, &hpa->central, size_min,
+	    size_goal);
+	malloc_mutex_unlock(tsdn, &hpa->mtx);
+	if (edata != NULL) {
+		edata_arena_ind_set(edata, shard->ind);
+		return edata;
+	}
+	/* No existing range can satisfy the request; try to grow. */
+	malloc_mutex_lock(tsdn, &hpa->grow_mtx);
+
+	/*
+	 * We could have raced with other grow attempts; re-check to see if we
+	 * did, and are now able to satisfy the request.
+	 */
+	malloc_mutex_lock(tsdn, &hpa->mtx);
+	edata = hpa_central_alloc_reuse(tsdn, &hpa->central, size_min,
+	    size_goal);
+	malloc_mutex_unlock(tsdn, &hpa->mtx);
+	if (edata != NULL) {
+		malloc_mutex_unlock(tsdn, &hpa->grow_mtx);
+		edata_arena_ind_set(edata, shard->ind);
+		return edata;
+	}
+
+	/*
+	 * No such luck. We've dropped mtx, so other allocations can proceed
+	 * while we allocate the new extent.  We know no one else will grow in
+	 * the meantime, though, since we still hold grow_mtx.
+	 */
+	size_t alloc_size;
+	pszind_t skip;
+
+	size_t hugepage_goal_min = HUGEPAGE_CEILING(size_goal);
+
+	err = geom_grow_size_prepare(&hpa->geom_grow, hugepage_goal_min,
+	    &alloc_size, &skip);
+	if (err) {
+		malloc_mutex_unlock(tsdn, &hpa->grow_mtx);
+		return NULL;
+	}
+	alloc_size = HUGEPAGE_CEILING(alloc_size);
+
+	/*
+	 * Eventually, we need to think about this more systematically, and in
+	 * terms of extent hooks.  For now, though, we know we only care about
+	 * overcommitting systems, and we're not going to purge much.
+	 */
+	bool commit = true;
+	void *addr = pages_map(NULL, alloc_size, HUGEPAGE, &commit);
+	if (addr == NULL) {
+		malloc_mutex_unlock(tsdn, &hpa->grow_mtx);
+		return NULL;
+	}
+	err = pages_huge(addr, alloc_size);
+	/*
+	 * Ignore this for now; even if the allocation fails, the address space
+	 * should still be usable.
+	 */
+	(void)err;
+
+	edata = edata_cache_get(tsdn, hpa->edata_cache);
+	if (edata == NULL) {
+		malloc_mutex_unlock(tsdn, &hpa->grow_mtx);
+		pages_unmap(addr, alloc_size);
+		return NULL;
+	}
+
+	edata_init(edata, hpa->ind, addr, alloc_size, /* slab */ false,
```

Comment:
why arena ind is hpa->ind? Should we use shard ind instead? (and remove L169?)

Reply:
Added some comments; the arena ind we use during growth should be the one associated with the global allocator. Once we've allocated a (probably much smaller) extent out of that large extent, it gets the arena index of the arena actually doing the allocation (rather than the index of the centralized allocator).